# Joining Data with Pandas

***

## Data Merging Basics

### Importing Libraries

In [1]:
# Import libarary
import pickle
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [3]:
# Load 1st dataset tax_owners

taxi_owners = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/taxi_owners.p", "rb")
)

In [4]:
# Load 2nd dataset taxi_vehicles

taxi_veh = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/taxi_veh.p", "rb")
)

---

### Showing Datasets

In [6]:
# 1st data
taxi_owners.head()

rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618

In [7]:
# 2nd data
taxi_veh.head()

vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO

#### Total Columns and Rows in these datasets

In [8]:
taxi_owners.shape

(3519, 5)

In taxi owners dataset there are 5 columns and 3,519 rows.

In [9]:
taxi_veh.shape

(3519, 6)

In taxi vehicles dataset there are 6 columns and 3,519 rows.

---

### Inner Join

**Merge taxi_owners with taxi_veh on the column vid, and save the result to taxi_own_veh**

In [10]:
# Merging two datasets

taxi_owners_veh = taxi_owners.merge(
    taxi_veh,       # merging 2nd dataset in 1st dataset
    on = 'vid'      # mergind id or same column 'vid'
)

In [12]:
# Printing out the new merged dataset
print(taxi_owners_veh.columns)

Index(['rid', 'vid', 'owner_x', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_y'],
      dtype='object')


In [13]:
# Showing merged dataset
taxi_owners_veh.head()

rid   vid         owner_x                 address    zip    make   model  \
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630  NISSAN  ALTIMA   
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659   HONDA     CRV   
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645  TOYOTA   CAMRY   
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   

   year fuel_type         owner_y  
0  2011    HYBRID  AGEAN TAXI LLC  
1  2014  GASOLINE    MANGIB CORP.  
2  2015  GASOLINE   FUNRIDE, INC.  
3  2014    HYBRID    ALQUSH CORP.  
4  2015  GASOLINE  EUNIFFORD INC.

**Set the left and right table suffixes for overlapping columns of the merge to _own and _veh, respectively.**

In [17]:
taxi_owners_veh = taxi_owners.merge(
    taxi_veh,
    on = 'vid',
    suffixes = ('_own', '_veh')
)

In [18]:
taxi_owners_veh.head()

rid   vid       owner_own                 address    zip    make   model  \
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630  NISSAN  ALTIMA   
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659   HONDA     CRV   
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645  TOYOTA   CAMRY   
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   

   year fuel_type       owner_veh  
0  2011    HYBRID  AGEAN TAXI LLC  
1  2014  GASOLINE    MANGIB CORP.  
2  2015  GASOLINE   FUNRIDE, INC.  
3  2014    HYBRID    ALQUSH CORP.  
4  2015  GASOLINE  EUNIFFORD INC.